bgpd/bgp_vty.c

```diff
@@ -3689,52 +3749,50 @@ DEFPY (no_bgp_bestpath_bw,
 	return CMD_SUCCESS;
 }
 
-/* "no bgp default ipv6-unicast". */
-DEFUN(no_bgp_default_ipv6_unicast, no_bgp_default_ipv6_unicast_cmd,
-      "no bgp default ipv6-unicast", NO_STR
+DEFPY(bgp_default_afi_safi, bgp_default_afi_safi_cmd,
+      "[no] bgp default <ipv4-unicast|"
+      "ipv4-multicast|"
+      "ipv4-vpn|"
+      "ipv4-flowspec|"
+      "ipv6-unicast|"
+      "ipv6-multicast|"
+      "ipv6-vpn|"
+      "ipv6-flowspec|"
+      "l2vpn-evpn>$afi_safi",
+      NO_STR
       "BGP specific commands\n"
       "Configure BGP defaults\n"
-      "Activate ipv6-unicast for a peer by default\n")
+      "Activate ipv4-unicast for a peer by default\n"
+      "Activate ipv4-multicast for a peer by default\n"
+      "Activate ipv4-vpn for a peer by default\n"
+      "Activate ipv4-flowspec for a peer by default\n"
+      "Activate ipv6-unicast for a peer by default\n"
+      "Activate ipv6-multicast for a peer by default\n"
+      "Activate ipv6-vpn for a peer by default\n"
+      "Activate ipv6-flowspec for a peer by default\n"
+      "Activate l2vpn-evpn for a peer by default\n")
 {
 	VTY_DECLVAR_CONTEXT(bgp, bgp);
-	UNSET_FLAG(bgp->flags, BGP_FLAG_DEFAULT_IPV6);
-	return CMD_SUCCESS;
-}
+	char afi_safi_str[strlen(afi_safi) + 1];
 
-DEFUN(bgp_default_ipv6_unicast, bgp_default_ipv6_unicast_cmd,
-      "bgp default ipv6-unicast",
-      "BGP specific commands\n"
-      "Configure BGP defaults\n"
-      "Activate ipv6-unicast for a peer by default\n")
-{
-	VTY_DECLVAR_CONTEXT(bgp, bgp);
-	SET_FLAG(bgp->flags, BGP_FLAG_DEFAULT_IPV6);
-	return CMD_SUCCESS;
-}
+	strcpy(afi_safi_str, afi_safi);
```

Comment:
strlcpy, please.

I would use `bgp default <afi> <safi>` instead of `bgp default <afi>-<safi>` at all. Or is here a real reason for that except that `bgp default ipv4-unicast` exists?

Reply:
yep, just saw the strlcpy warning so I'm fixing that now.

yes, that's the exact reason I'm maintaining the hyphen.  I also think two separate args is cleaner, but ipv4-unicast has existed for a while and I didn't want to break backwards compatibility or introduce inconsistency in the UI.

Reply:
Just did a force push that resolves this (and moves from strtok -> strtok_r).